### PR TITLE
Playground

### PR DESCRIPTION
**Pull request by JamesIlling**
Creating PR for work done

### CODE REVIEW

File: Domain/EmailSender.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+using MailKit;
+using MimeKit;
+using Domain;
+using Domain.Interfaces;
+
+namespace Domain
+{
+    public class EmailSender: IEmailSender
+    {
+
```
Comment (JamesIlling) [inline Domain/EmailSender.cs]: remove the additional whitespace

File: Domain/EmailSender.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                client.Connect("smtp.gmail.com", 587, false);
+
+                // Note: only needed if the SMTP server requires authentication
+                client.Authenticate("CwPoolChampion", "NgBYNK4A5QgN2ss8");
+
+                client.Send(message);
+                client.Disconnect(true);
+            }
+        }
+
```
Comment (JamesIlling) [inline Domain/EmailSender.cs:41]: What happens on an invalid email address?

File: Domain/EmailSender.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
```
Comment (JamesIlling) [inline Domain/EmailSender.cs:42]: Does it support HTML emails?

File: Domain/EmailSender.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                client.ServerCertificateValidationCallback = (s, c, h, e) => true;
+
+                client.Connect("smtp.gmail.com", 587, false);
+
+                // Note: only needed if the SMTP server requires authentication
+                client.Authenticate("CwPoolChampion", "NgBYNK4A5QgN2ss8");
+
+                client.Send(message);
+                client.Disconnect(true);
+            }
```
Comment (JamesIlling) [inline Domain/EmailSender.cs:39]: Where are the assertions ? this does not actually test anything

File: DomainTest/ScoreTest.cs [hunk 5]
```diff
@@ -1,10 +1,25 @@
...
+    [TestFixture]
     public class ScoreTest
     {
-        public ScoreTest()
+
+        [Test]
+        [Category("Unit")]
+        public void CanCreateNewScore()
         {
+            var player1 = new Player("Daniel");
```
Comment (JamesIlling) [inline DomainTest/ScoreTest.cs:17]: Why is this here, it isnt part of the action and isnt needed for the assertions

File: DomainTest/RankingTest.cs [hunk 6]
```diff
@@ -63,5 +112,105 @@ public void CalculateScores_RankingsContainsAllPlayers()
...
+            score1.Player.ShouldBe("Daniel");
+            score1.Points.ShouldBe(1);
+        }
+
+
+        [Test]
+        [Category("Unit")]
+        public void CanCreateNewRanking()
+        {
+            Ranking newRanking = new Ranking();
```
Comment (JamesIlling) [inline DomainTest/RankingTest.cs:209]: This test is a little redundant, are all you are checking is an exception is not thrown from the constructor, however the constructor should be as small as possible and not throw an exception.

File: DomainTest/RankingTest.cs [hunk 7]
```diff
@@ -63,5 +112,105 @@ public void CalculateScores_RankingsContainsAllPlayers()
...
+
+
+            var game1 = player2.CreateGame(player1, player2);
+
+            player1.SetGameResults(game1, player1);
+            player1.ConfirmGame(game1);
+            var score1 = new Score(player1.Name);
+            var ranking = new Ranking();
+            
+            ranking.actionOnWinnerAfterGame(game1, score1);
```
Comment (JamesIlling) [inline DomainTest/RankingTest.cs]: The capitalisation on this method is wrong should be ActionOnWinnerAfterGame. Why does this action exist? all we need is the calculate scores.

File: Domain/Interfaces/IRanking.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Domain.Interfaces
+{
+    public interface IRanking
+    {
+        IEnumerable<Score> CalculateScores(List<Game> games);
+        List<Score> GetRanking(List<Score> scores);
```
Comment (JamesIlling) [inline Domain/Interfaces/IRanking.cs]: What does this method do? all we need is the CalculateScores

File: Domain/Player.cs [hunk 9]
```diff
@@ -22,8 +22,44 @@ public Player(string name, string emailAddress)
...
+
+            if (game.Player1.Name == this.Name || game.Player2.Name == this.Name)
+            {
+                game.Winner = player;
+                game.EmailSender.SendEmail(playerToBeSentConfirmationEmail);//players should have like a "pending confirmation inbox" which will be automatically loaded first thing when logging in. in case there is no items pending first thing loaded will be the creation of game form. 
+            }
+           
+        }
+
+        public Player DecideWhoIsToBeSentConfirmationEmail(Game game)
```
Comment (JamesIlling) [inline Domain/Player.cs:46]: Why is this part of the player? This will be part of a service

File: Domain/Player.cs [hunk 10]
```diff
@@ -22,8 +22,44 @@ public Player(string name, string emailAddress)
...
+        public void ConfirmGame(Game game)
+        {
+            if((game.Player1.Name == this.Name || game.Player2.Name == this.Name) && this.Name != game.Creator.Name)
+            {
+                game.IsConfirmed = true;//here we will have to wait for the confirmation coming back from player that has been sent a link to the confirmation modal.
 
+            }
+        }
+
+        public void SetGameResults(Game game, Player player)
```
Comment (JamesIlling) [inline Domain/Player.cs:34]: Why is this part of the player? it should be part of the game

File: Domain/Player.cs [hunk 11]
```diff
@@ -22,8 +22,44 @@ public Player(string name, string emailAddress)
             EmailAdress = emailAddress;
         }
 
+        public void ConfirmGame(Game game)
```
Comment (JamesIlling) [inline Domain/Player.cs:25]: Why is this part of a player?

File: Domain/Ranking.cs [hunk 12]
```diff
@@ -1,12 +1,95 @@
...
+
+        public List<Score> GetRanking(List<Score> scores)
+        {
+             scores.Sort((a, b) => b.Points.CompareTo(a.Points));
+            return scores;
+        }
+
+
+
+        public void applyActionOnPlayersAfterGame(Game game)
```
Comment (JamesIlling) [inline Domain/Ranking.cs]: The capitalization of this method is wrong, also why is this public

File: Domain/Ranking.cs [hunk 13]
```diff
@@ -1,12 +1,95 @@
...
+        {
+            if(score.Player == game.Winner.Name && game.IsConfirmed == true)
+            {
+                score.Points += 1;
+            }
+        }
+
+   
+
+        public void insertAllPlayerNamesInScoresArray(List<Game> games)
```
Comment (JamesIlling) [inline Domain/Ranking.cs]: The capitalization of this method is wrong, also why is this public

File: Domain/Ranking.cs [hunk 14]
```diff
@@ -1,12 +1,95 @@
...
+            {
+                Score score = new Score(name);
+                score.Points = 0;
+                scores.Add(score);
+            }
         }
+
+
+
+        public IEnumerable<string> getAllPlayerNames(List<Game> games)
```
Comment (JamesIlling) [inline Domain/Ranking.cs]: The capitalization of this method is wrong, also why is this public

File: Domain/Ranking.cs [hunk 15]
```diff
@@ -1,12 +1,95 @@
...
+            foreach (Game game in games)
+            {
+                applyActionOnPlayersAfterGame(game);
+            }
+
+            GetRanking(scores);
+            return scores;
+        }
+
+        public List<Score> GetRanking(List<Score> scores)
```
Comment (JamesIlling) [inline Domain/Ranking.cs]: why is this not part of calculate scores? Why is this public

File: Domain/Ranking.cs [hunk 16]
```diff
@@ -1,12 +1,95 @@
 ﻿using System.Collections.Generic;
+using Domain.Interfaces;
 
+
+//how can i make a method be able to return 2 different things. if there is an IF statement?
+//IEnumerable, List, Array? why can i not use linq to order by with list. what is the difference between all of them?
+//how to do it so that only a particular class can change the properties of another class and how to do it so only one specific class can call another class constructor
+//when you compare "==" two objects that are meant to be the same, what exactly makes it the same? is there like an internal form of ID or something.
+//what is the name for the flow of how and when the methods are being called.?
 namespace Domain
```
Comment (JamesIlling) [inline Domain/Ranking.cs:10]: You cant make it return two things the closest is the type or null.
yes you can, LINQ returns an IQueryable which you would need to use the .ToList() method on to get a list back
You should never control access at that level, you only have public, internal,protected, and private
it depends on the type of the object
there isnt a na,e

File: Domain/Game.cs [hunk 17]
```diff
@@ -8,16 +9,20 @@ public class Game
         public Player Player1 { get; set; }
         public Player Player2 { get; set; }
         public Player Winner { get; set; }
-
+        public Boolean IsConfirmed { get; internal set; }
+        public DateTime TimeStamp { get; }
+        public Player Creator { get; set; }
+        public IEmailSender EmailSender { get; }
```
Comment (JamesIlling) [inline Domain/Game.cs]: Why is sending an email part of the game?

File: Domain/EmailSender.cs (hunk 4 above)
Comment (DanielLaraSanchez) [reply Domain/EmailSender.cs:39]: i know, i was just using the test to run the method to see if the email was actually being sent. now that i have added the validation i can actually test that. is there a way to test if the email has been sent?

File: Domain/EmailSender.cs (hunk 2 above)
Comment (DanielLaraSanchez) [reply Domain/EmailSender.cs:41]: true! i have changed that

File: Domain/EmailSender.cs (hunk 1 above)
Comment (DanielLaraSanchez) [reply Domain/EmailSender.cs]: ok

File: DomainTest/ScoreTest.cs (hunk 5 above)
Comment (DanielLaraSanchez) [reply DomainTest/ScoreTest.cs:17]: because the game takes two players. and i need the game in order to create a score.

File: Domain/Ranking.cs (hunk 14 above)
Comment (DanielLaraSanchez) [reply Domain/Ranking.cs]: ok, that has been changed

File: Domain/Ranking.cs (hunk 15 above)
Comment (DanielLaraSanchez) [reply Domain/Ranking.cs]: yes, you are right.. i took the "s" a bit too far

File: Domain/Ranking.cs (hunk 14 above)
Comment (DanielLaraSanchez) [reply Domain/Ranking.cs]: how can i test it otherwise?

File: Domain/Ranking.cs (hunk 13 above)
Comment (DanielLaraSanchez) [reply Domain/Ranking.cs]: how can i test it otherwise?

Comment (DanielLaraSanchez) [reply Domain/Ranking.cs]: capitalisation has been changed

File: Domain/Ranking.cs (hunk 14 above)
Comment (DanielLaraSanchez) [reply Domain/Ranking.cs]: capitalisation has been changed

File: Domain/Game.cs (hunk 17 above)
Comment (DanielLaraSanchez) [reply Domain/Game.cs]: true, that has been changed
